### PR TITLE
Add implementation for AssemblyName::EscapedCodebase

### PR DESCRIPTION
**Pull request by rahku**

### CODE REVIEW

**Comment by rahku** [comment]
On full desktop clr we use windows OS apis to escape uri string. Adding implementation of the code from mono codebase so that we have implementation for other OS's as well.

**Comment by rahku** [comment]
PTAL @jkotas @weshaggard

**Comment by rahku** [comment]
PTAL @kouvel

File: src/mscorlib/src/System/Reflection/AssemblyName.cs [hunk 1]
```diff
@@ -533,7 +534,83 @@ internal string GetNameWithPublicKey()
...
+		{
+			if (character > 255) {
+				throw new ArgumentOutOfRangeException ("character");
+			}
+			
+			return "%" + hexUpperChars [((character & 0xf0) >> 4)] 
+			           + hexUpperChars [((character & 0x0f))];
+		}
+        
+        private static readonly string hexUpperChars = "0123456789ABCDEF";
```
Comment (jkotas) [inline src/mscorlib/src/System/Reflection/AssemblyName.cs]: readonly -> const

**Comment by jkotas** [comment]
> implementation of the code from mono codebase

Licensing notices that need to be included with this?

File: src/mscorlib/src/System/Reflection/AssemblyName.cs [hunk 2]
```diff
@@ -533,7 +571,83 @@ internal string GetNameWithPublicKey()
         private extern byte[] nGetPublicKeyToken();
     
         [System.Security.SecurityCritical]  // auto-generated
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        static internal extern String EscapeCodeBase(String codeBase);
+        static internal String EscapeCodeBase(String codebase)
+        {
+            return EscapeString(codebase, false, true, true);
```
Comment (kouvel) [inline src/mscorlib/src/System/Reflection/AssemblyName.cs]: It looks like the desktop framework is not escaping #, [, or ]. It's also not escaping the % when what follows makes it not a hex escape. For instance, "%hi" or "%1g"

File: src/mscorlib/src/System/Reflection/AssemblyName.cs [hunk 3]
```diff
@@ -533,7 +571,83 @@ internal string GetNameWithPublicKey()
...
+        }
+
+        static internal String EscapeString(String str, bool escapeReserved, bool escapeHex, bool escapeBrackets)
```
Comment (kouvel) [inline src/mscorlib/src/System/Reflection/AssemblyName.cs]: It looks like most of this functionality is already implemented in CoreFX UriHelper.EscapeString, which Uri.EscapeUriString uses. Although Uri.EscapeUriString seems to re-escape hex-escaped values, UriHelper.EscapeString can be configured as needed. Would it be better to call that through reflection instead, or to move that implementation into CoreLib?

File: src/mscorlib/src/System/Reflection/AssemblyName.cs [hunk 4]
```diff
@@ -533,7 +571,83 @@ internal string GetNameWithPublicKey()
...
+		{
+			if (character > 255) {
+				throw new ArgumentOutOfRangeException ("character");
+			}
+			
+			return "%" + hexUpperChars [((character & 0xf0) >> 4)] 
+			           + hexUpperChars [((character & 0x0f))];
+		}
+        
+        private static const string hexUpperChars = "0123456789ABCDEF";
```
Comment (kouvel) [inline src/mscorlib/src/System/Reflection/AssemblyName.cs]: static const -> const?

File: src/mscorlib/src/System/Reflection/AssemblyName.cs (hunk 3 above)
Comment (jkotas) [inline src/mscorlib/src/System/Reflection/AssemblyName.cs]: It is just a few methods - I would create copy of the corefx impl in in CoreLib.

File: src/mscorlib/src/System/Reflection/AssemblyName.cs (hunk 2 above)
Comment (rahku) [inline src/mscorlib/src/System/Reflection/AssemblyName.cs]: I tried on desktop clr #, [, & ] seems to be converted to corresponding escape sequences and matches what mono does. You are correct about %hi though.

File: src/mscorlib/src/System/Reflection/AssemblyName.cs (hunk 3 above)
Comment (rahku) [inline src/mscorlib/src/System/Reflection/AssemblyName.cs]: I don't think this api is used much and if we should try and make it compatible with desktop. This implementation works fine for xamarin and so I believe should be fine for .net core. @jkotas what do you think.

File: src/mscorlib/src/System/Reflection/AssemblyName.cs (hunk 2 above)
Comment (kouvel) [inline src/mscorlib/src/System/Reflection/AssemblyName.cs]: Here's the code I tried:
```c#
        var an = new AssemblyName();
        an.CodeBase = "#[]";
        Console.WriteLine(an.EscapedCodeBase);
```
I'm getting the output `#[]`. Is there something I'm doing differently?

**Comment by rahku** [comment]
PTAL I now use the escaping implementation from corefx repo instead of mono. This now matches what System.Uri.EscapeUriString api does.

**Comment by kouvel** [comment]
Why have this match the EscapeUriString behavior? It should be fairly easy to have this match the desktop escaping behavior by passing in different arguments. As for the FormatException, I feel that it is appropriate even though in desktop I guess (haven't tried) it would just escape the invalid surrogate pairs too. But if we want to match that too, we could pass in another argument to turn off the exception and escape (in the first case) or skip (in the second case when decoding yields zero bytes, not sure how that could happen, if at all).

**Comment by rahku** [comment]
https://github.com/dotnet/coreclr/pull/8072#discussion_r87493853 @jkotas suggested that we should have the same behavior throughout net core stack and I agree with that. This is also what mono does.

**Comment by kouvel** [comment]
Ah ok, I thought it was just in reference to the implementation and not the behavior. If it's ok to change the behavior to match Uri, that's great. LGTM as well

**Comment by rahku** [comment]
@dotnet-bot test this please

**Comment by rahku** [comment]
@dotnet-bot test Windows_NT x64 Debug Build and Test

**Comment by jkotas** [comment]
@dotnet-bot test OSX x64 Checked Build and Test please

**Pull request merged**